test/WhenOrderTest.php

```diff
@@ -0,0 +1,30 @@
+<?php declare(strict_types = 1);
+
+namespace Amp\Test;
+
+use Amp\Deferred;
+
+class WhenOrderTest extends \PHPUnit_Framework_TestCase {
+    public function testWhenOrder() {
+        $this->expectOutputString("1234");
+
+        $deferred = new Deferred;
+        $promise = $deferred->promise();
+
+        $promise->when(function () use ($promise) {
+            $promise->when(function () {
```

Comment:
I had not considered that. We have to define when a promise is resolved. I guess it should be _before_ the callbacks are invoked.